src/packet/grammar.ts

```diff
@@ -0,0 +1,121 @@
+  }
+
+  // One-Pass Signed Message:
+  //    One-Pass Signature Packet, OpenPGP Message, Corresponding Signature Packet.
+  if (tagList.findIndex(tag => tag === enums.packet.onePassSignature) === 0) {
+    const correspondingSigPacketIndex = util.findLastIndex(tagList, tag => tag === enums.packet.signature);
+    if (correspondingSigPacketIndex !== tagList.length - 1 && !acceptPartial) {
+      return false;
+    }
+    return isValidOpenPGPMessage(tagList.slice(1, correspondingSigPacketIndex < 0 ? undefined : correspondingSigPacketIndex), acceptPartial);
+  }
+
+  return false;
+};
+
+const isUnknownPacketTag = (tag: number): tag is enums.packet => {
+  try {
+    enums.read(enums.packet, tag);
+    return false;
+  } catch (e) {
+    return true;
+  }
+};
+
+/**
+ * Implements grammar checks based on https://www.rfc-editor.org/rfc/rfc9580.html#section-10.3 .
+ * @param notNormalizedList - list of packet tags to validate
+ * @param acceptPartial - whether the list of tags corresponds to a partially-parsed message
+ * @returns whether the list of tags is valid
+ */
+const isValidOpenPGPMessage = (
+  notNormalizedList: number[] /** might have unknown tags */,
+  acceptPartial: boolean
+): boolean => {
+  // Take care of packet tags that can appear anywhere in the sequence:
+  // 1. A Marker packet (Section 5.8) can appear anywhere in the sequence.
+  // 2. An implementation MUST be able to process Padding packets anywhere else in an OpenPGP stream so that future revisions of this document may specify further locations for padding.
+  // 3. An unknown non-critical packet MUST be ignored (criticality its enforced on parsing).
+  const normalizedList: enums.packet[] = notNormalizedList.filter(tag => (
+    tag !== enums.packet.marker &&
+    tag !== enums.packet.padding &&
+      !isUnknownPacketTag(tag)
+  ));
+
+  return isValidLiteralMessage(normalizedList, acceptPartial) ||
+    isValidCompressedMessage(normalizedList, acceptPartial) ||
+    isValidEncryptedMessage(normalizedList, acceptPartial) ||
+    isValidSignedMessage(normalizedList, acceptPartial);
+};
+
+/**
+ * If `delayReporting === false`, the grammar validator throws as soon as an invalid packet sequence is detected during parsing.
+ * This setting MUST NOT be used when parsing unauthenticated decrypted data, to avoid instantiating decryption oracles.
+ *  Passing `delayReporting === true` allows checking the grammar validity in an async manner, by
+ * only reporting the validity status after parsing is done (i.e. and authentication is expected to
+ * have been enstablished)
+ * @returns `true` on successful grammar validation; if `delayReporting` is set, `null` is returned
+ *   if validation is still pending (partial parsing, waiting for authentication to be confirmed).
+ * @throws on grammar error, provided `config.enforceGrammar` is enabled.
+ */
+export const getMessageGrammarValidator = ({ delayReporting }: { delayReporting: boolean }) => {
+  let logged = false;
+
```

Comment:
This is a bit nitpicky but the `@returns` and `@throws` apply to the function below, not this one:

```suggestion
 */
export const getMessageGrammarValidator = ({ delayReporting }: { delayReporting: boolean }) => {
  let logged = false;

  /**
   * @returns `true` on successful grammar validation; if `delayReporting` is set, `null` is returned
   *   if validation is still pending (partial parsing, waiting for authentication to be confirmed).
   * @throws on grammar error, provided `config.enforceGrammar` is enabled.
   */
```